### PR TITLE
Journalist can login to create articles

### PR DESCRIPTION
**Pull request by swardbladpaulo**
[Journalist can log in to create an article -Client 2](https://www.pivotaltracker.com/story/show/176055823)
### User Story
```
As a journalist,
In order to write articles
I would like to be able to log in
```
## Changes proposed in this pull request:

* Writing a test to see if journalist can login
* Creating Authentication component
* Creating Header component
* Adding rootReducer and configureStore in State folder
* Creating CreateArticleForm component.
* Create a test for journalist can create articles.
* Add styling to App.jsx, CreateArticleForm.jsx

## What have we learned working on this feature:
*  How to use functional components.
* How to use React-Redux.

## Packages/Gems added
* Semantic UI
* Axios
* React-Redux

## Screenshots (Client)
Screenshot from 2020-12-13 20-39-39

### CODE REVIEW

File: src/components/CreateArticleForm.jsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        },
+        {
+          headers: headers,
+        }
+      );
+    } catch (err) {
+      console.error(err);
+      alert("Something went wrong!");
+    }
+  };
```
Comment (emtalen) [inline src/components/CreateArticleForm.jsx]: refactor this logic to a module

File: cypress/integration/journalistCanCreateArticles.feature.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      url: "http://localhost:3000/api/auth/sign_in",
+      response: "fixture:journalist_can_login.json",
+      headers: {
+        uid: "journalist@mail.com",
+      },
+    });
+    cy.route({
+      method: "POST",
+      url: "http://localhost:3000/api/articles",
+      response: {},
```
Comment (emtalen) [inline cypress/integration/journalistCanCreateArticles.feature.js]: You expect to see a message but you are stubbing out an empty response, from where is the message coming?

File: src/components/CreateArticleForm.jsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            content: content,
+          },
+        },
+        {
+          headers: headers,
+        }
+      );
+    } catch (err) {
+      console.error(err);
+      alert("Something went wrong!");
```
Comment (emtalen) [inline src/components/CreateArticleForm.jsx]: use the error messages that you get from the backend

File: src/components/CreateArticleForm.jsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  const [subtitle, setSubtitle] = useState();
+  const [content, setContent] = useState();
+
+  let headers = useSelector((state) => state.currentUser);
+  const saveArticle = async (e) => {
+    e.preventDefault();
+    setTitle(e.target.title.value);
+    setSubtitle(e.target.input_sub_title.value);
+    setContent(e.target.input_content.value);
+    setMessage("Your article was created");
```
Comment (emtalen) [inline src/components/CreateArticleForm.jsx]: use the status message you get form the backend

**Review by emtalen — CHANGES_REQUESTED** [review]

File: src/components/CreateArticleForm.jsx (hunk 1 above)
Comment (KyunginNa) [reply src/components/CreateArticleForm.jsx]: We extracted it to createArticle module!

File: cypress/integration/journalistCanCreateArticles.feature.js (hunk 2 above)
Comment (KyunginNa) [reply cypress/integration/journalistCanCreateArticles.feature.js]: Added message in response

File: src/components/CreateArticleForm.jsx (hunk 3 above)
Comment (KyunginNa) [reply src/components/CreateArticleForm.jsx]: Yes we set response.data.message instead

File: src/components/CreateArticleForm.jsx (hunk 4 above)
Comment (KyunginNa) [reply src/components/CreateArticleForm.jsx]: We set response.data.message for this one too

File: cypress/integration/journalistCanCreateArticles.feature.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      );
+      cy.get('[data-cy="create-article-button"]').click();
+    });
+    cy.get('[data-cy="response-message"]').should(
+      "contain",
+      "Your article was successfully created!"
+    );
+  });
+
+  it("unauthorized", () => {
```
Comment (emtalen) [inline cypress/integration/journalistCanCreateArticles.feature.js]: What happens if I log in as a user and not a journalist? 
Also, I'm missing the sad path for what happens if I try to create an article without a title for example?

File: cypress/integration/journalistCanLogin.feature.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    cy.get("[data-cy='flash-message']").should("contain", "You are logged in");
+  });
+
+  it("sad path: unsuccessfully", () => {
+    cy.route({
+      method: "POST",
+      url: "http://localhost:3000/api/auth/sign_in",
+      status: "401",
+      response: {
+        errors: ["Login failed, please try again"],
```
Comment (emtalen) [inline cypress/integration/journalistCanLogin.feature.js]: this is not the error you get from the backend

File: cypress/integration/journalistCanLogin.feature.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        success: false,
+      },
+    });
+
+    cy.get("[data-cy='login-form']").within(() => {
+      cy.get("[data-cy='email']").type("journalist@mail.com");
+      cy.get("[data-cy='password']").type("wrongpassword");
+      cy.get("[data-cy='submit-btn']").contains("Submit").click();
+    });
+    cy.get("[data-cy='header-user-email']").contains("You're not logged in.");
```
Comment (emtalen) [inline cypress/integration/journalistCanLogin.feature.js:76]: I want some message telling me want went wrong with my login attempt

File: src/components/Authentication.jsx [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+const Authentication = () => {
+  const dispatch = useDispatch();
+  const performAuthentication = async (event) => {
+    event.preventDefault();
+    let credentials = {
+      email: event.target.email.value,
+      password: event.target.password.value,
+    };
+    let response = await axios.post("/auth/sign_in", credentials);
```
Comment (emtalen) [inline src/components/Authentication.jsx]: you are still not using J-tockauth here...

File: src/components/CreateArticleForm.jsx (hunk 1 above)
Comment (emtalen) [reply src/components/CreateArticleForm.jsx]: there is no need to set the inputs in state, you have access to them by state, please refactor that.

**Review by emtalen — CHANGES_REQUESTED** [review]

File: cypress/integration/journalistCanCreateArticles.feature.js (hunk 5 above)
Comment (rithreaksa) [reply cypress/integration/journalistCanCreateArticles.feature.js]: We fixed this! Thanks :D

File: cypress/integration/journalistCanLogin.feature.js (hunk 6 above)
Comment (rithreaksa) [reply cypress/integration/journalistCanLogin.feature.js]: We fixed this one too.

File: cypress/integration/journalistCanLogin.feature.js (hunk 7 above)
Comment (rithreaksa) [reply cypress/integration/journalistCanLogin.feature.js:76]: This is fixed too.

File: src/components/Authentication.jsx (hunk 8 above)
Comment (rithreaksa) [reply src/components/Authentication.jsx]: We refactored to use J-tockauth now.

File: src/modules/CreateArticle.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        {
+          headers: headers,
+        }
+      );
+      dispatch({
+        type: "SET_ARTICLE_MESSAGE",
+        payload: response.data.message,
+      });
+    } catch (err) {
+      console.log(err);
```
Comment (emtalen) [inline src/modules/CreateArticle.js]: You need to set an error handler here, there are a common use case to get an error message when making a post request, if the title isn't filled and so on

**Review by emtalen — CHANGES_REQUESTED** [review]

File: src/modules/CreateArticle.js (hunk 9 above)
Comment (KyunginNa) [reply src/modules/CreateArticle.js]: Thanks for help! We fixed it :)

**Review by emtalen — APPROVED** [review]

**Pull request merged**